File: services/auth-service-py/src/infrastructure/db/repositories/oracle_sede_repository.py
from typing import Optional

from src.domain.entities.sede import Sede
from src.domain.repositories.i_sede_repository import ISedeRepository
from src.infrastructure.db.connection import get_pool

_SELECT = """
    SELECT
        ID           AS id,
        CODIGO       AS codigo,
        NOMBRE       AS nombre,
        REGION_ID    AS region_id,
        REGION       AS region,
        DEPARTAMENTO AS departamento,
        PROVINCIA    AS provincia,
        DIRECCION    AS direccion,
        TELEFONO     AS telefono,
        ACTIVO       AS activo
    FROM SEC_SEDE
"""


def _row_to_sede(row) -> Sede:
    return Sede(
        id=row[0],
        codigo=row[1],
        nombre=row[2],
        region_id=row[3],
        region=row[4],
        departamento=row[5],
        provincia=row[6],
        direccion=row[7],
        telefono=row[8],
        activo=row[9],
    )


class OracleSedeRepository(ISedeRepository):

    async def find_all_active(self) -> list[Sede]:
        pool = get_pool()
        async with pool.acquire() as conn:
            async with conn.cursor() as cur:
                await cur.execute(
                    f"{_SELECT} WHERE ACTIVO = 1 ORDER BY NOMBRE"
                )
                rows = await cur.fetchall()
                return [_row_to_sede(r) for r in rows]

    async def find_by_id(self, sede_id: int) -> Optional[Sede]:
        pool = get_pool()
        async with pool.acquire() as conn:
            async with conn.cursor() as cur:
                await cur.execute(
                    f"{_SELECT} WHERE ID = :sede_id",
                    {"sede_id": sede_id},
                )
                row = await cur.fetchone()
                return _row_to_sede(row) if row else None

    async def find_by_codigo(self, codigo: str) -> Optional[Sede]:
        pool = get_pool()
        async with pool.acquire() as conn:
            async with conn.cursor() as cur:
                await cur.execute(
                    f"{_SELECT} WHERE CODIGO = :codigo",
                    {"codigo": codigo},
                )
                row = await cur.fetchone()
                return _row_to_sede(row) if row else None

File: services/auth-service/src/infrastructure/db/repositories/OracleSedeRepository.ts
import { Repository } from 'typeorm';
import { SedeEntity } from '../entities/SedeEntity';
import { ISedeRepository } from '../../../domain/repositories/ISedeRepository';
import { Sede } from '../../../domain/entities/Sede';

/**
 * Repositorio Oracle — usa raw SQL para evitar el problema de
 * TypeORM que genera identificadores entre comillas dobles (case-sensitive
 * en Oracle), lo que causa ORA-00904.
 * Oracle devuelve los nombres de columna en MAYÚSCULAS en raw queries.
 */
export class OracleSedeRepository implements ISedeRepository {
  constructor(private readonly repo: Repository<SedeEntity>) {}

  private readonly SELECT_SEDE = `
    SELECT id, codigo, nombre, region_id, region, departamento, provincia, activo
    FROM   SEC_SEDE
  `;

  private rowToSede(row: any): Sede {
    return new Sede(
      row.ID,
      row.CODIGO,
      row.NOMBRE,
      row.REGION_ID,
      row.REGION,
      row.DEPARTAMENTO,
      row.PROVINCIA,
      row.ACTIVO === 1,
    );
  }

  async findAll(): Promise<Sede[]> {
    const rows = await this.repo.query(
      `${this.SELECT_SEDE} WHERE activo = 1 ORDER BY nombre ASC`
    );
    return rows.map((r: any) => this.rowToSede(r));
  }

  async findById(id: number): Promise<Sede | null> {
    const rows = await this.repo.query(
      `${this.SELECT_SEDE} WHERE id = :1`, [id]
    );
    return rows.length > 0 ? this.rowToSede(rows[0]) : null;
  }

  async findByCodigo(codigo: string): Promise<Sede | null> {
    const rows = await this.repo.query(
      `${this.SELECT_SEDE} WHERE codigo = :1`, [codigo]
    );
    return rows.length > 0 ? this.rowToSede(rows[0]) : null;
  }

  async findByRegion(regionId: number): Promise<Sede[]> {
    const rows = await this.repo.query(
      `${this.SELECT_SEDE} WHERE region_id = :1 AND activo = 1 ORDER BY nombre ASC`,
      [regionId]
    );
    return rows.map((r: any) => this.rowToSede(r));
  }
}
